RavenDB-8775 - local license update

grisha-kotler opened the pull request.

grisha-kotler committed:
RavenDB-8775 - add the ability to update the license locally (env variables, settings.json and license path)

grisha-kotler committed:
RavenDB-8775 - update the license when it's updated in the cluster

ayende commented on src/Raven.Server/Commercial/LicenseHelper.cs:
```diff
@@ -0,0 +1,230 @@
+                if (RavenConfiguration.EnvironmentVariableLicenseString != null)
+                {
+                    // update the environment variable string value
+                    UpdateEnvironmentVariableLicenseString(newLicense, rsaParameters);
```
This usually won't be persistent, though

ayende commented on src/Raven.Server/Commercial/LicenseHelper.cs:
```diff
@@ -0,0 +1,230 @@
+                }
+
+                // update the license string in settings.json
+                if (TryUpdatingSettingsJson(newLicense, rsaParameters))
```
Is this also the priority order for reading the license?

ayende commented on src/Raven.Server/Commercial/LicenseHelper.cs:
```diff
@@ -0,0 +1,230 @@
+            if (licenseString == null)
+            {
+                if (Logger.IsInfoEnabled)
+                    Logger.Info($"Failed to access the environment variable: `{RavenConfiguration.EnvironmentVariableLicenseString}`");
```
Won't this also trigger if we don't have the env var setup at all?

ayende commented on src/Raven.Server/Commercial/LicenseHelper.cs:
```diff
@@ -0,0 +1,230 @@
+                return;
+            }
+
+            var oldLicense = DeserializeLicense(licenseString);
```
shouldn't this be on a try/catch?

ayende commented on src/Raven.Server/Commercial/LicenseHelper.cs:
```diff
@@ -0,0 +1,230 @@
+                return;
+
+            var newLicenseString = JsonConvert.SerializeObject(newLicense);
+            Environment.SetEnvironmentVariable(RavenConfiguration.EnvironmentVariableLicenseString, newLicenseString, preferredTarget);
```
This is non persistent set.
For that matter, I'm not sure that there is something that we _can_ do properly here.
Consider a docker instance, which may have the env setup externally completely.

Shouldn't we keep a local copy of the license inside the server store instead as well?

ayende commented on src/Raven.Server/Commercial/LicenseHelper.cs:
```diff
@@ -0,0 +1,230 @@
+                return true;
+            }
+
+            if (oldLicense.Id != newLicense.Id)
```
Why does this matter?

grisha-kotler committed:
RavenDB-8775 - prevent concurrent update of the local license

grisha-kotler committed:
RavenDB-8775 - change the order of trying to update the license:
- license string - settings.json
- license string - environment variable
- license path
